sources/api/apiserver/src/bin/apiserver.rs

```diff
@@ -131,15 +139,50 @@ async fn run() -> Result<()> {
     let args = parse_args(env::args());
 
     // SimpleLogger will send errors to stderr and anything less to stdout.
-    SimpleLogger::init(args.log_level, LogConfig::default())
-        .context(error::Logger)?;
+    SimpleLogger::init(args.log_level, LogConfig::default()).context(error::Logger)?;
 
     // Make sure the datastore exists
     ensure!(
         Path::new(&args.datastore_path).exists(),
         error::NonexistentDatastore
     );
 
+    // We need to handle some actions asynchronously without holding up API clients, like invoking
+    // thar-be-settings to apply changes to the system, so we fork them off and don't wait for
+    // their status or output at that point.  Instead, when we receive a SIGCHLD signal telling us
+    // that a child process has exited, we wait for it in this thread so that it doesn't become a
+    // zombie process.
+    let mut signals = Signals::new(&[SIGCHLD]).context(error::Signal)?;
+    thread::spawn(move || {
+        for signal in signals.forever() {
+            match signal {
+                SIGCHLD => {
+                    // We loop as long as we have children to reap.  If you only wait for one
+                    // process per signal, you'll still accumulate zombies (in periods of fast
+                    // activity) if this signal handler is interrupted by another signal.
+                    loop {
+                        if let Ok(status) = waitpid(Pid::from_raw(-1), Some(WaitPidFlag::WNOHANG)) {
+                            // You can get a CHLD signal in other scenarios, even if a child is
+                            // still running; we only want to continue if we're making progress
+                            // with reaping exited children.  Otherwise, stop and wait for another
+                            // signal.
+                            if !matches!(status, WaitStatus::Exited(..)) {
+                                break;
+                            }
+                        } else {
+                            // We can't handle errors in any fancy way here because we can't take
+                            // many actions in a signal handler, like printing, and we don't have
+                            // much recourse anyway.  If there are still children in the next
+                            // iteration, we'll try again.
```

Comment:
Looking at the implementation of `signal_hook` it seems like what it does is set up a signal handler where the action is writing to a pipe. The iterator runs in a normal context and returns items from that pipe, so this function isn't the actual signal handler.

Relative to the comments:
* it shouldn't be possible for another signal to interrupt us, since we're using `WNOHANG` in our system call
* it should be safe to print whatever we want

The code looks fine to me; we're just not giving `signal_hook` enough credit for all the heavy lifting it's doing.

Reply:
Interesting. I definitely had zombies stacking up if I didn't have the loop, and I would expect the 1:1 signal->wait to work if it can't be interrupted...